MTD geometry: material budget validation

fabiocos opened the pull request:
#### PR description:

This PR updates the Validation/Geometry package by adding a class for the evaluation of the MTD material budget. A macro providing plots is also added. The clang-format cleaning has been applied to the whole package.

#### PR validation:

Results of a 10k neutrinos run on scenario D38:

[Mtd_Materials_l_vs_eta.pdf](https://github.com/cms-sw/cmssw/files/3195412/Mtd_Materials_l_vs_eta.pdf)
[Mtd_Materials_x_vs_eta.pdf](https://github.com/cms-sw/cmssw/files/3195413/Mtd_Materials_x_vs_eta.pdf)

fabiocos committed:
scram b code-format-all

fabiocos committed:
MTD material budget validation class and configuration

fabiocos committed:
Add macro to produce plots

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26830/9857

- This PR adds an extra 64KB to repository

cmsbuild commented:
A new Pull Request was created by  @fabiocos (Fabio Cossutti) for master.

It involves the following packages:

Validation/Geometry


@andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @cmsbuild, @jfernan2, @fioriNTU, @civanch can you please review it and eventually sign? Thanks.
@vargasa, @rishabhCMS, @rovere, @rbartek this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fabiocos commented:
please test

testing really nothing here...

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e1afd9c92d9605db67b3304d2ee4de8c75e20930
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/328/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/328/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211772
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211437
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

jfernan2 commented:
+1

fabiocos committed:
Update material table to include services for scenario D41

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26830/9876

- This PR adds an extra 64KB to repository

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26830 was updated. @andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @cmsbuild, @jfernan2, @fioriNTU, @civanch can you please check and sign again.

cmsbuild commented:
+1
Tested at: 19b221b40df43d1b9fbdcae458727c29c9aff0aa
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/351/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/351/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3206828
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3206493
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

fabiocos commented:
@pmeridian @abenagli please let me know whether the format looks ok, or you would prefer to change something. I've run it for D35, D38 and D41 with 1M input neutrinos.

cvuosalo commented on Validation/Geometry/src/MaterialBudgetMtdHistos.cc:
```diff
@@ -0,0 +1,296 @@
+      "1630", "MB prof Eta  Phi [Other];#eta;#varphi;#lambda/#lambda_{0}", 250, -5., 5., 180, -3.1416, 3.1416));
+  hmgr->addHisto2(new TH2F("1631", "Eta vs Phi [Other]", 501, -5., 5., 180, -3.1416, 3.1416));
+  hmgr->addHistoProf1(new TProfile("1640", "MB prof R [Other];R [mm];#lambda/#lambda_{0}", 200, 0., 2000.));
+
```
Our coding guidelines recommend using constants for all these magic numbers: 5, 180, 250, 501, etc.

fabiocos replied:
@cvuosalo you are correct, this was a quick copy of the tracker class. But as it is new it makes sense to update and get it clean asap, it makes also my work easier. Please see the new version

cvuosalo commented:
+1

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26830/9969

- This PR adds an extra 64KB to repository

cmsbuild commented:
Pull request #26830 was updated. @andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @cmsbuild, @jfernan2, @fioriNTU, @civanch can you please check and sign again.

fabiocos committed:
Update macros to comply with CMS guidelines for publication plots

fabiocos committed:
Redefine material categories, address C. Vuosalo comment on constants definition

jfernan2 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 94535b8cefb0a2d4b653a1314e788425c5f72d76
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/514/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/514/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3206856
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3206519
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26830/9992

- This PR adds an extra 68KB to repository

cmsbuild commented:
Pull request #26830 was updated. @andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @jfernan2, @fioriNTU, @civanch can you please check and sign again.

fabiocos commented:
@cvuosalo @jfernan2 I do not plan further updates for this PR, I will have one more in a forthcoming PR for a small update of the MTD geometry scenario. Here are the plots coming out from this setup:

[Mtd_Materials_x_vs_eta.pdf](https://github.com/cms-sw/cmssw/files/3220144/Mtd_Materials_x_vs_eta.pdf)
[Mtd_Materials_l_vs_eta.pdf](https://github.com/cms-sw/cmssw/files/3220145/Mtd_Materials_l_vs_eta.pdf)

As discussed with @pmeridian we have now the CMS publication suggested presentation (with the CMS_lumi and tdrstyle macros slightly adjusted compared to the version available in https://twiki.cern.ch/twiki/bin/view/CMS/Internal/FigGuidelines ). And I have addressed the comment of @cvuosalo about the definition of constants (aka "magic numbers"), which is indeed correct (although their meaning in this case was quite trivial).

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9555e5998797dbd086b68f6d86968ac212666f82
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/518/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/518/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3206856
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3206519
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

jfernan2 commented:
+1

ianna commented on Validation/Geometry/interface/MaterialBudget.h:
```diff
@@ -21,37 +21,34 @@ class EndOfTrack;
-  
+
 private:
+  MaterialBudget(const MaterialBudget&) = delete;                   // stop default
```
@fabiocos - always make deleted methods public because it is an announcement to the public users that they cannot be used.

ianna commented on Validation/Geometry/interface/MaterialBudget.h:
```diff
@@ -21,37 +21,34 @@ class EndOfTrack;
-  std::vector<int>              constituents, detLevels,regionTypes,stackOrder;
-  std::vector<double>           etaRegions, boundaries;
+
+  std::vector<std::string> detTypes, detNames;
```
@fabiocos - naming rule #2.9: private members should be named as ```member_ ```

ianna commented on Validation/Geometry/interface/MaterialBudgetForward.h:
```diff
@@ -21,41 +21,38 @@ class EndOfTrack;
-  std::vector<double>           stepLen, radLen, intLen;
-  double                        eta, phi, stepT;
+private:
+  MaterialBudgetForward(const MaterialBudgetForward &) = delete;                   // stop default
```
@fabiocos - public, please

ianna commented on Validation/Geometry/interface/MaterialBudgetHcal.h:
```diff
@@ -15,32 +15,29 @@ class BeginOfTrack;
-  
+
 private:
+  MaterialBudgetHcal(const MaterialBudgetHcal&) = delete;                   // stop default
```
@fabiocos - public, please

ianna commented on Validation/Geometry/interface/MaterialBudgetHcal.h:
```diff
@@ -15,32 +15,29 @@ class BeginOfTrack;
   bool stopAfter(const G4Step*);
-  
-  MaterialBudgetHcalHistos*   theHistoHcal;
+
```
@fabiocos - rule No2.9

ianna reviewed (changes requested):
Please, make sure that the rule No 2.9 is implemented in naming private members.

ianna commented:
-1

fabiocos commented:
@ianna your comments are technically correct, but please notice that they have nothing to do with the real content of this PR, as they affect classes that enters only through the clang format cleaning in the first commit that I did to help the core migration and avoid interferences. I think that we may profit of the fact that we touch the whole CMSSW code base to catch old issues, and I have fixed some here and there, but we should keep their fix separate from the cleaning in order not to slow down the latter, unless it is the cleaning itself to create issues (which is not the case here). Otherwise we should likely rewrite most of out code base. Anyway your suggestions are relatively small, I will add them.

fabiocos committed:
Address comments by Y. Osborne about coding style

cmsbuild commented:
The code-checks are being triggered in jenkins.

fabiocos commented:
@ianna although not related with PR but for the clang-format cleaning, I have addressed a couple of issues you have raised by

- promoting to public "delete" methods;

- applying rule 2.9 were needed, as it says _A trailing “_” is the preferred method to distinguish a data member from the getter method_ . I have addressed the cases where there are variables and methods with the same name (not those pointed by you).

-

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26830/10067

- This PR adds an extra 148KB to repository

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26830 was updated. @andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @cmsbuild, @jfernan2, @fioriNTU, @civanch can you please check and sign again.

cmsbuild commented:
+1
Tested at: 5898b9bf7cac1c1555d11b25453e06fda9dfa4a0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/607/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-54bd0f/607/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3210678
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3210343
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

fabiocos commented:
@ianna @cvuosalo any further comment? or can we move forward?

cvuosalo commented:
+1

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
